tests/integration-tests.test.ts

```diff
@@ -409,4 +409,109 @@ describe('integration tests', () => {
+            changeReqFn(req);
+            next();
+         });
+
+         helloRouteHandler2 = spy((req: Request, resp: Response): void => {
+            hello2Req = _.pick(req, ...reqProps);
+            resp.send('hello');
+         });
+
+         goodbyeRouteHandler = spy((req: Request, resp: Response): void => {
+            goodbyeReq = _.pick(req, ...reqProps);
+            resp.send('goodbye');
+         });
+
+         router.get('/hello', helloRouteHandler)
+            .get('/goodbye', goodbyeRouteHandler)
+            .get('/hello', helloRouteHandler2);
+
+         app.addSubRouter('/path', router);
+
+         testOutcome('GET', '/path/hello', expectedResponse);
+
+         return {
+            hello: helloRouteHandler,
+            getHelloReq: () => { return helloReq; },
+            hello2: helloRouteHandler2,
+            getHello2Req: () => { return hello2Req; },
+            goodbye: goodbyeRouteHandler,
+            getGoodbyeReq: () => { return goodbyeReq; },
+         };
+      }
+
+      it('redirects to a different route handler when req.url is changed', () => {
+         const handlers = testRoutesWithRedirect();
+
+         assert.calledOnce(handlers.hello);
+         assert.calledOnce(handlers.goodbye);
+         assert.notCalled(handlers.hello2);
+         assert.callOrder(handlers.hello, handlers.goodbye);
+      });
+
+      it('continues executing routes in the current chain before redirecting to the next match', () => {
+         const router = new Router();
+
+         let h1, h2Redirects, h3, h4ShouldBeSkipped, g1ShouldBeSkipped, g2Ends;
+
+         h1 = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+
+         h2Redirects = spy((req: Request, _resp: Response, next: NextCallback): void => {
+            req.url = '/goodbye';
+            next();
+         });
+
+         h3 = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         h4ShouldBeSkipped = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         g1ShouldBeSkipped = spy((_req: Request, _resp: Response, next: NextCallback): void => { next(); });
+         g2Ends = spy((_req: Request, resp: Response): void => { resp.send('goodbye'); });
+
+         router.get('/goodbye', g1ShouldBeSkipped)
+            .get('/hello', h1, h2Redirects, h3)
+            .get('/hello', h4ShouldBeSkipped)
+            .get('/goodbye', g2Ends);
+
+         app.addSubRouter('/path', router);
+
+         testOutcome('GET', '/path/hello', 'goodbye');
+
+         assert.notCalled(g1ShouldBeSkipped);
+         assert.calledOnce(h1);
+         assert.calledOnce(h2Redirects);
+         assert.calledOnce(h3);
+         assert.notCalled(h4ShouldBeSkipped);
+         assert.calledOnce(g2Ends);
+         assert.callOrder(h1, h2Redirects, h3, g2Ends);
+      });
+
+      it('ignores query string params when `request.url` changes to a URL with query params', () => {
```

Comment:
I feel like we need to document this behavior. Maybe the property documentation on `url` needs to explain that you can modify it mid-request to cause an internal route-handling redirect, and explicitly state what happens to query string parameters during a redirect (that apparently they are ignored).